rustPlatform.importCargoLock: add support for git dependencies that use workspace inheritance

winterqt opened the pull request:
###### Description of changes

Rust 1.64.0 added support for workspace inheritance, which allows
for crates to inherit values such as dependency version constraints or
package metadata information from their workspaces [0].

This works by having workspace members specify a value as a table, with
`workspace` set to true. Thus, supporting this in importCargoLock is as
simple as walking the crate's Cargo.toml, replacing inherited values
with their workspace counterpart.

This is also what a forthcoming Cargo release will do for `cargo vendor` [1],
but we can get ahead of it ;)

[0]: https://blog.rust-lang.org/2022/09/22/Rust-1.64.0.html#cargo-improvements-workspace-inheritance-and-multi-target-builds
[1]: https://github.com/rust-lang/cargo/pull/11414

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

winterqt committed:
rustPlatform.importCargoLock: add support for git dependencies that use workspace inheritance

Rust 1.64.0 added support for workspace inheritance, which allows
for crates to inherit values such as dependency version constraints or
package metadata information from their workspaces [0].

This works by having workspace members specify a value as a table, with
`workspace` set to true. Thus, supporting this in importCargoLock is as
simple as walking the crate's Cargo.toml, replacing inherited values
with their workspace counterpart.

This is also what a forthcoming Cargo release will do for `cargo vendor` [1],
but we can get ahead of it ;)

[0]: https://blog.rust-lang.org/2022/09/22/Rust-1.64.0.html#cargo-improvements-workspace-inheritance-and-multi-target-builds
[1]: https://github.com/rust-lang/cargo/pull/11414

winterqt committed:
matrix-conduit: use importCargoLock instead of hacky patch

As `cargo vendor` currently doesn't support workspace inherited manifest
values, a patch that replaced the values manually was used as a workaround.

Now that importCargoLock has support for this, and we're probably going
to move towards it anyways, let's switch to it now, as there's a clear
benefit either way.

This also switches to using our copy of SQLite instead of the vendored
one included in `libsqlite3-sys` -- because it's preferred, but also
because it causes the build to fail, now that dependencies are read-only.

winterqt commented:
I wanted to include bumping Ruff here, but ran into https://github.com/charliermarsh/ruff/issues/3039. If it gets fixed before this is merged, I'll add it.

figsoda commented on pkgs/build-support/rust/test/import-cargo-lock/git-dependency-workspace-inheritance/default.nix:
```diff
@@ -0,0 +1,7 @@
+{ replaceWorkspaceValues, runCommand }:
```
instead of testing the behavior of `replaceWorskpaceValues`, what if we just test `cargoLock.lockFile` directly?

```toml
[dependencies.rustpython-compiler-core]
package = "rustpython-compiler-core"
git = "https://github.com/RustPython/RustPython"
```

winterqt replied:
We could do that, but I also kind of prefer the simplistic nature of the current test, as it's tiny + adequately tests what the tool does.

Either works -- let's wait for others to weigh in.

figsoda commented on pkgs/build-support/rust/replace-workspace-values.py:
```diff
@@ -0,0 +1,43 @@
+            if isinstance(workspace_copy, str):
+                replaced["version"] = workspace_copy
+            else:
+                replaced.update(workspace_copy)
```
> [`features`](https://doc.rust-lang.org/cargo/reference/features.html) declared in this table are additive with the `features` from `[dependencies]`

We should probably handle this

https://doc.rust-lang.org/cargo/reference/workspaces.html#the-dependencies-table

winterqt replied:
Agreed, done.

figsoda commented on pkgs/build-support/rust/replace-workspace-values.py:
```diff
@@ -0,0 +1,57 @@
+            else:
+                replaced.update(workspace_copy)
+
+            merged_features = []
+
+            if crate_features is not None:
+                merged_features.extend(crate_features)
+
+            if (workspace_features := workspace_copy.get("features")) is not None:
+                merged_features.extend(workspace_features)
```
I think this should do the same thing, and we can move this logic (and whatever `dict.fromkeys` does I'm assuming) after (or before if we mutate workspace_copy) `replaced.update(workspace_copy)`
```suggestion
            merged_features = set((crate_features or []) + (workspace_copy.get("features") or []))
```

winterqt replied:
You know, when writing that, I was thinking to myself "if only Python had `or`." 😄

figsoda commented on pkgs/build-support/rust/replace-workspace-values.py:
```diff
@@ -0,0 +1,51 @@
+def replace_dependencies(root):
+    for key in ["dependencies", "dev-dependencies", "build-dependencies"]:
+        if key in root:
+            for k, v in root[key].items():
```
```suggestion
            for k in root[key].keys():
```

winterqt replied:
Oops, result of a refactor and I didn't change that line. Surprised flake8 didn't warn about it...

winterqt committed:
ruff: 0.0.244 -> 0.0.248

https://github.com/charliermarsh/ruff/releases/tag/v0.0.245
https://github.com/charliermarsh/ruff/releases/tag/v0.0.246
https://github.com/charliermarsh/ruff/releases/tag/v0.0.247
https://github.com/charliermarsh/ruff/releases/tag/v0.0.248

figsoda commented on pkgs/build-support/rust/replace-workspace-values.py:
```diff
@@ -0,0 +1,84 @@
+    if "workspace" in crate_manifest:
+        return
+
+    orig_crate_manifest = copy.deepcopy(crate_manifest)
```
is this really worth it?

figsoda replied:
As an alternative, we should probably have a global that is set to true [here](https://github.com/NixOS/nixpkgs/blob/0cc4830f14578ece6a3e0c1e543785570e26f2e4/pkgs/build-support/rust/replace-workspace-values.py#L23), unconditionally `deepcopy`ing the whole table doesn't look like optimization to me, if that's what you were going for

winterqt replied:
Fair, not sure what I was thinking there 😕

winterqt replied:
(I ended up not using a global, but a bool returned from each function and ORed.)

figsoda reviewed (approved):
Can't speak much about the changes to `matrix-conduit`, would appreciate a review from one of the maintainers.

The `importCargoLock` and `ruff` changes lgtm, would prefer if tests are done using `buildRustPackage` instead but this is up for discussion

figsoda commented:
Result of `nixpkgs-review pr 217232` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>flare-floss</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>matrix-conduit</li>
    <li>python310Packages.python-flirt</li>
    <li>python311Packages.python-flirt</li>
    <li>ruff</li>
    <li>sapling</li>
  </ul>
</details>

winterqt commented:
All of those failures are unrelated.

Darwin: dependency being missed after the libiconv changes (I presume, haven't looked at history to confirm).
Linux (and I presume Darwin, after fixing the above): dependency version mismatch?

figsoda commented:
Yep I was just posting this so other people don't have to rebuild these

winterqt commented:
Fixing these now...

wischli commented:
Could you make an estimation when a consumer would be able to make use of this fix inside any nix (pre) release?

figsoda commented:
This is already in master and the unstable branches, if you are using stable, then 23.05

wischli commented:
> This is already in master and the unstable branches, if you are using stable, then 23.05

Thanks for the quick reply. Any chance it will be backported to 22.11?

figsoda commented:
@wischli can you test #223804?

wischli commented:
> @wischli can you test #223804?

Unfortunately, the error still persists

```
       > Caused by:
       >   error inheriting `version` from workspace root manifest's `workspace.package.version`
```

This is the relevant feature where we encounter this: https://github.com/centrifuge/centrifuge-chain/pull/1241
One of our dependencies ([Polkadot](https://github.com/paritytech/polkadot)) switched to workspace inheritance.

figsoda commented:
This fix only applies to `importCargoLock`, the fix for `fetchCargoTarball` will come with the rust 1.68 update, which will not be backported to 22.11

```patch
-             cargoSha256 = "sha256-BkSVd08QksOvyIRO1bVpiOnWf440r+dLR1A3UiND7IM=";
+             cargoLock = {
+               lockFile = ./Cargo.lock;
+               allowBuiltinFetchGit = true;
+             };
```

wischli commented:
Thanks for sharing this information with us. Looking forward to the release which supports workspace inheritance in `fetchCargoTarball`.
